src/pgen/progenitor.cpp

```diff
@@ -0,0 +1,180 @@
+  auto Srr_adm= progenitor_pkg->Param<Spiner::DataBox>("Srr_adm");
+
+
+  auto mass_density_dev= progenitor_pkg->Param<Spiner::DataBox>("mass_density_dev");
+  auto Ye_dev= progenitor_pkg->Param<Spiner::DataBox>("Ye_dev");
+  auto velocity_dev= progenitor_pkg->Param<Spiner::DataBox>("velocity_dev");
+  auto pressure_dev = progenitor_pkg->Param<Spiner::DataBox>("pressure_dev");
+
+
+  
+  //MonopoleGR
+  static bool monopole_initialized = false;
+  if (!monopole_initialized) {
+  auto monopole_pkg = pmb->packages.Get("monopole_gr");
+  auto matter = monopole_pkg->Param<MonopoleGR::Matter_t>("matter");
+  auto matter_h = monopole_pkg->Param<MonopoleGR::Matter_host_t>("matter_h");
+  auto npoints = monopole_pkg->Param<int>("npoints");
+  auto rad = monopole_pkg->Param<MonopoleGR::Radius>("radius");
+  
+  
+  
+  for (int i = 0; i < npoints; ++i){
+    matter_h(MonopoleGR::Matter::RHO,i)=adm_density.interpToReal(rad.x(i)); 
+    matter_h(MonopoleGR::Matter::J_R,i)=adm_momentum.interpToReal(rad.x(i));
+    matter_h(MonopoleGR::Matter::trcS,i)=S_adm.interpToReal(rad.x(i));
+    matter_h(MonopoleGR::Matter::Srr,i)=Srr_adm.interpToReal(rad.x(i));
+  }
+  Kokkos::deep_copy(matter,matter_h);
+  MonopoleGR::IntegrateHypersurface(monopole_pkg.get());
+  MonopoleGR::LinearSolveForAlpha(monopole_pkg.get());
+  MonopoleGR::SpacetimeToDevice(monopole_pkg.get());
+  MonopoleGR::DumpToTxt("phoebus-metric-setup.dat",monopole_pkg.get());
+  monopole_initialized = true;
+  } // if statement
+  
+  // Primitive Variables
+  using Transformation_t = Geometry::SphericalToCartesian;
+  auto const &geom_pkg = pmb->packages.Get("geometry");
+  auto transform = Geometry::GetTransformation<Transformation_t>(geom_pkg.get());
+  pmb->par_for(
+      "Phoebus::ProblemGenerator::homologous", kb.s, kb.e, jb.s, jb.e, ib.s, ib.e,
+      KOKKOS_LAMBDA(const int k, const int j, const int i) {
+      Real r;
+      const Real x1 = coords.Xc<1>(k, j, i);
+      const Real x2 = coords.Xc<2>(k, j, i);
+      const Real x3 = coords.Xc<3>(k, j, i);
+      Real C[3];
+      Real s2c[3][3], c2s[3][3];
+
+      if (is_monopole_sph) {
+          r = std::abs(x1);
+        }
+      else { // Cartesian
+          r = std::sqrt(x1 * x1 + x2 * x2 + x3 * x3);
+	  transform(x1, x2, x3, C, c2s, s2c);
+        }
+	
+        Real lambda[2];
+        if (iye > 0) {
+          v(iye, k, j, i) = Ye_dev.interpToReal(r);
+          lambda[0] = v(iye, k, j, i);
+        }
+	
+	const Real u = phoebus::energy_from_rho_P(eos, mass_density_dev.interpToReal(r),pressure_dev.interpToReal(r),emin,emax,lambda[0]);
+	const Real sie = u/mass_density_dev.interpToReal(r);
+        const Real T = eos.TemperatureFromDensityInternalEnergy(mass_density_dev.interpToReal(r), sie, lambda);
+
+	Real vel_vec_in[3]={0};
+	Real vel_vec_out[3]={0};
+	vel_vec_in[0] = velocity_dev.interpToReal(r);
+	if (is_monopole_cart){
+	  for ( int i=0; i<3; ++i){
+	    for (int j=0; j<3; ++j){
```

Comment:
I'm not sure if this is more or less clear, but Phoebus provides a macro for 3D loops. You could replace these two loops with
```C++
SPACELOOP2(i, j) {
```
and then you would need one fewer close `}` below as well.